src/Conversion/ONNXToTOSA/Math/Gemm.cpp

```diff
@@ -0,0 +1,208 @@
+      std::vector<float> elements = {};
+      for (int i = 0; i < cformat[0]; ++i)
+        elements.push_back(0.0F);
+      C = mlir::tosa::getConstTensor<float>(rewriter, op, elements, cformat).value();
+    }
+
+    rewriter.replaceOpWithNewOp<tosa::FullyConnectedOp>(op, resultType, A, B, C);
+    return success();
+  }
+
+  /// The GEMM can be described as a FullyConnected operator
+  /// Y = AB^T + C if we perform a transpose on B only with
+  /// alpha and beta factors set to 1.
+  /// The constraints on rank were taken from the torch implementation check
+  /// Input A must be of rank 2 (input)
+  /// Input B must be of rank 2 (weights)
+  /// Input C must be of rank 1 (bias)
+  static bool checkLegalTosaFCOp(ONNXGemmOp op, OpAdaptor adaptor) {
+    Value A = op.A();
+    Value B = op.B();
+    Value C = op.C();
+
+    auto AType = A.getType().cast<RankedTensorType>();
+    auto BType = B.getType().cast<RankedTensorType>();
+
+    // If C is present, it can only be of rank 1
+    if (C.getType().isa<RankedTensorType>() &&
+        C.getType().cast<RankedTensorType>().getRank() != 1) {
+      return false;
+    }
+    // Input tensor must be of rank 2
+    // Weights must also be of rank 2
+    if (AType.getRank() != 2 || BType.getRank() != 2) {
+      return false;
+    }
+    // Both alpha and beta must be 1
+    if ((adaptor.alpha().convertToFloat() != 1.0F) ||
+        (adaptor.beta().convertToFloat() != 1.0F)) {
+      return false;
+    }
+    // Only Transpose B must be enabled
+    if (adaptor.transA() != 0 || adaptor.transB() != 1) {
+      return false;
+    }
+    return true;
+  }
+  LogicalResult matchAndRewrite(ONNXGemmOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    Location loc = op->getLoc();
+    Value A = op.A();
+    Value B = op.B();
+    Value C = op.C();
+    int64_t transA = adaptor.transA();
+    int64_t transB = adaptor.transB();
+    FloatAttr alpha = adaptor.alphaAttr();
+    FloatAttr beta = adaptor.betaAttr();
+    auto AType = A.getType().cast<RankedTensorType>();
+    auto BType = B.getType().cast<RankedTensorType>();
+    auto shapeA = A.getType().dyn_cast<ShapedType>().getShape();
+    auto shapeB = B.getType().dyn_cast<ShapedType>().getShape();
+    auto resultType = getTypeConverter()->convertType(op.getResult().getType()).cast<TensorType>();
+    
+    // C is optional, if it's not there, we need to be aware of it for later computations
+    bool isCPresent = C.getType().isa<TensorType>();
+    // If it's not present, ONNX creates a NoValue operator. Remove it.
+    if (!isCPresent) {
+      rewriter.eraseOp(C.getDefiningOp());
+    }
+
+    // If legal, create a FullyConnected operator instead
+    if (checkLegalTosaFCOp(op, adaptor)) {
+      return rewriteToTosaFC(op, adaptor, rewriter);
+    }
+    
+    // ONNX gives 2d matrix as input and expect a 2d output. TOSA expects everything to be 3D. As such, there
```

Comment:
I think the more common terminology here (pytorch etc.) is less 2D vs 3D, but rather that TOSA provides a batch matmul. Coming from ONNX the batch is set to one by definition, and not because it is unknown (it isn't unknown - we have a single batch to process).